# Homework Assignment - Lesson 7: String Manipulation and Date/Time Data

**Student Name:** Nick Eisenheim

**Student ID:** qtz882

**Date Submitted:** [October 12 2025]

**Due Date:** October 12 2025

---

## Objective

Master string manipulation with `stringr` and date/time operations with `lubridate` for real-world business data cleaning and analysis.

## Learning Goals

By completing this assignment, you will:
- Clean and standardize messy text data using `stringr` functions
- Parse and manipulate dates using `lubridate` functions
- Extract information from text and dates for business insights
- Combine string and date operations for customer segmentation
- Create business-ready reports from raw data

## Instructions

- Complete all tasks in this notebook
- Write your code in the designated TODO sections
- Use the pipe operator (`%>%`) wherever possible
- Add comments explaining your logic
- Run all cells to verify your code works
- Answer all reflection questions

## Datasets

You will work with three CSV files:
- `customer_feedback.csv` - Customer reviews with messy text
- `transaction_log.csv` - Transaction records with dates
- `product_catalog.csv` - Product descriptions needing standardization

---

## Part 1: Data Import and Initial Exploration

**Business Context:** Before cleaning data, you must understand its structure and quality issues.

**Your Tasks:**
1. Load required packages (`tidyverse` and `lubridate`)
2. Import all three CSV files from the `data/` directory
3. Examine the structure and identify data quality issues
4. Display sample rows to understand the data

In [2]:
# Task 1.1: Load Required Packages
# TODO: Load tidyverse (includes stringr)
library(tidyverse)
library(lubridate)
setwd("/workspaces/assignment-1-nheisenheim/data")
# TODO: Load lubridate


cat("✅ Packages loaded successfully!\n")

✅ Packages loaded successfully!


In [3]:

# Task 1.2: Import Datasets
# TODO: Import customer_feedback.csv into a variable called 'feedback'
feedback <- read.csv("customer_feedback.csv")

# TODO: Import transaction_log.csv into a variable called 'transactions'
transactions <- read.csv("transaction_log.csv")

# TODO: Import product_catalog.csv into a variable called 'products'
products <- read.csv("product_catalog.csv")

cat("✅ Data imported successfully!\n")
cat("Feedback rows:", nrow(feedback), "\n")
cat("Transaction rows:", nrow(transactions), "\n")
cat("Product rows:", nrow(products), "\n")

✅ Data imported successfully!
Feedback rows: 100 
Transaction rows: 150 
Product rows: 75 


In [4]:
# Task 1.3: Initial Data Exploration

cat("=== CUSTOMER FEEDBACK DATA ===\n")
# TODO: Display structure of feedback using str()
str(feedback)

# TODO: Display first 5 rows of feedback
print(head(feedback,5))

cat("\n=== TRANSACTION DATA ===\n")
# TODO: Display structure of transactions
str(transactions)

# TODO: Display first 5 rows of transactions
print(head(transactions,5))

cat("\n=== PRODUCT CATALOG DATA ===\n")
# TODO: Display structure of products
str(products)

# TODO: Display first 5 rows of products
print(head(products,5))


=== CUSTOMER FEEDBACK DATA ===
'data.frame':	100 obs. of  6 variables:
 $ FeedbackID   : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID   : int  12 40 34 1 47 13 13 37 49 23 ...
 $ Customer_Name: chr  "Bob Wilson" "   sarah johnson   " "jane smith" "JANE SMITH" ...
 $ Feedback_Text: chr  "Highly recommend this item" "   Excellent service   " "Poor quality control" "average product, nothing special" ...
 $ Contact_Info : chr  "bob.wilson@test.org" "555-123-4567" "jane_smith@company.com" "jane_smith@company.com" ...
 $ Feedback_Date: chr  "2024-02-23" "2024-01-21" "2023-09-02" "2023-08-21" ...


  FeedbackID CustomerID       Customer_Name                    Feedback_Text
1          1         12          Bob Wilson       Highly recommend this item
2          2         40    sarah johnson                Excellent service   
3          3         34          jane smith             Poor quality control
4          4          1          JANE SMITH average product, nothing special
5          5         47       michael brown      AMAZING customer support!!!
            Contact_Info Feedback_Date
1    bob.wilson@test.org    2024-02-23
2           555-123-4567    2024-01-21
3 jane_smith@company.com    2023-09-02
4 jane_smith@company.com    2023-08-21
5           555-123-4567    2023-04-24

=== TRANSACTION DATA ===
'data.frame':	150 obs. of  5 variables:
 $ LogID               : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID          : int  26 21 12 6 32 27 31 30 31 13 ...
 $ Transaction_DateTime: chr  "4/5/24 14:30" "3/15/241 14:30" "3/15/24 14:30" "3/20/24 9:15" ...
 $ Amount              

## Part 2: String Cleaning and Standardization

**Business Context:** Product names and feedback text often have inconsistent formatting that prevents accurate analysis.

**Your Tasks:**
1. Clean product names (remove extra spaces, standardize case)
2. Standardize product categories
3. Clean customer feedback text
4. Extract customer names from feedback

**Key Functions:** `str_trim()`, `str_squish()`, `str_to_lower()`, `str_to_upper()`, `str_to_title()`

In [5]:
# Task 2.1: Clean Product Names
# TODO: Create a new column 'product_name_clean' that:
#   - Removes leading/trailing whitespace using str_trim()
#   - Converts to Title Case using str_to_title()

products_clean <- products %>%
  mutate(
    # Your code here:
    product_name_clean = str_trim(Product_Description),
    product_name_clean = str_to_title(product_name_clean),
    product_name_clean = str_replace_all(product_name_clean, "\\s+", " ")


  )

# Display before and after
cat("Product Name Cleaning Results:\n")
products_clean %>%
  select(Product_Description, product_name_clean) %>%
  head(10) %>%
  print()

Product Name Cleaning Results:
                         Product_Description
1  Apple iPhone 14 Pro - 128GB - Space Black
2             samsung galaxy s23 ultra 256gb
3  Apple iPhone 14 Pro - 128GB - Space Black
4  Apple iPhone 14 Pro - 128GB - Space Black
5             samsung galaxy s23 ultra 256gb
6  Apple iPhone 14 Pro - 128GB - Space Black
7   DELL XPS 13 Laptop - Intel i7 - 16GB RAM
8         hp envy printer - wireless - color
9   Nike Air Max 270 - Size 10 - Black/White
10         LG 55" 4K Smart TV - OLED Display
                          product_name_clean
1  Apple Iphone 14 Pro - 128gb - Space Black
2             Samsung Galaxy S23 Ultra 256gb
3  Apple Iphone 14 Pro - 128gb - Space Black
4  Apple Iphone 14 Pro - 128gb - Space Black
5             Samsung Galaxy S23 Ultra 256gb
6  Apple Iphone 14 Pro - 128gb - Space Black
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram
8         Hp Envy Printer - Wireless - Color
9   Nike Air Max 270 - Size 10 - Black/White
10         Lg 55" 4k Sma

In [6]:
# Task 2.2: Standardize Product Categories
# TODO: Create a new column 'category_clean' that:
#   - Converts category to Title Case
#   - Removes any extra whitespace

products_clean <- products_clean %>%
  mutate(
    # Your code here:
    category_clean = str_trim(Category),
    category_clean = str_to_title(category_clean),
    category_clean = str_replace_all(category_clean, "\\s+", " ")
  )

# Show unique categories before and after
cat("Original categories:\n")
print(unique(products$category))

cat("\nCleaned categories:\n")
print(unique(products_clean$category_clean))

Original categories:
NULL

Cleaned categories:
[1] "Tv"          "Audio"       "Shoes"       "Electronics" "Computers"  


In [7]:
# Task 2.3: Clean Customer Feedback Text
# TODO: Create a new column 'feedback_clean' that:
#   - Converts text to lowercase using str_to_lower()
#   - Removes extra whitespace using str_squish()

feedback_clean <- feedback %>%
  mutate(
    # Your code here:
    feedback_clean = str_to_lower(Feedback_Text),
    feedback_clean = str_squish(feedback_clean),
    feedback_clean = str_replace_all(feedback_clean, "[^a-z0-9\\s]", "")
  )

# Display sample
cat("Feedback Cleaning Sample:\n")
feedback_clean %>%
  select(Feedback_Text, feedback_clean) %>%
  head(5) %>%
  print()

Feedback Cleaning Sample:
                     Feedback_Text                  feedback_clean
1       Highly recommend this item      highly recommend this item
2             Excellent service                  excellent service
3             Poor quality control            poor quality control
4 average product, nothing special average product nothing special
5      AMAZING customer support!!!        amazing customer support


## Part 3: Pattern Detection and Extraction

**Business Context:** Identifying products with specific features and extracting specifications helps with inventory management and marketing.

**Your Tasks:**
1. Identify products with specific keywords (wireless, premium, gaming)
2. Extract numerical specifications from product names
3. Detect sentiment words in customer feedback
4. Extract email addresses from feedback

**Key Functions:** `str_detect()`, `str_extract()`, `str_count()`

In [8]:
# Task 3.1: Detect Product Features
# TODO: Create three new columns:
#   - is_wireless: TRUE if product name contains "wireless" (case-insensitive)
#   - is_premium: TRUE if product name contains "pro", "premium", or "deluxe"
#   - is_gaming: TRUE if product name contains "gaming" or "gamer"
# Hint: Use str_detect() with str_to_lower() for case-insensitive matching
# Hint: Use | (pipe) in regex for OR conditions

products_clean <- products_clean %>%
  mutate(
    # Your code here:
    is_wireless = str_detect(str_to_lower(product_name_clean), "wireless"),
    is_premium = str_detect(str_to_lower(product_name_clean), "pro|premium|deluxe"),
    is_gaming = str_detect(str_to_lower(product_name_clean), "gaming|gamer")
  )

# Display results
cat("Product Feature Detection:\n")
products_clean %>%
  select(product_name_clean, is_wireless, is_premium, is_gaming) %>%
  head(10) %>%
  print()

# Summary statistics
cat("\nFeature Summary:\n")
cat("Wireless products:", sum(products_clean$is_wireless), "\n")
cat("Premium products:", sum(products_clean$is_premium), "\n")
cat("Gaming products:", sum(products_clean$is_gaming), "\n")

Product Feature Detection:
                          product_name_clean is_wireless is_premium is_gaming
1  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
2             Samsung Galaxy S23 Ultra 256gb       FALSE      FALSE     FALSE
3  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
4  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
5             Samsung Galaxy S23 Ultra 256gb       FALSE      FALSE     FALSE
6  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram       FALSE      FALSE     FALSE
8         Hp Envy Printer - Wireless - Color        TRUE      FALSE     FALSE
9   Nike Air Max 270 - Size 10 - Black/White       FALSE      FALSE     FALSE
10         Lg 55" 4k Smart Tv - Oled Display       FALSE      FALSE     FALSE

Feature Summary:
Wireless products: 17 
Premium products: 13 
Gaming products: 0 


In [9]:
# Task 3.2: Extract Product Specifications
# TODO: Create a new column 'size_number' that extracts the first number from product_name
# Hint: Use str_extract() with pattern "\\\\d+" to match one or more digits

products_clean <- products_clean %>%
  mutate(
    # Your code here:
    size_number = str_extract(product_name_clean, "\\\\d+")
  )

# Display products with extracted sizes
cat("Extracted Product Specifications:\n")
products_clean %>%
  filter(!is.na(size_number)) %>%
  select(product_name_clean, size_number) %>%
  head(10) %>%
  print()

Extracted Product Specifications:
[1] product_name_clean size_number       
<0 rows> (or 0-length row.names)


In [10]:
# Task 3.3: Simple Sentiment Analysis
# TODO: Create three new columns:
#   - positive_words: count of positive words ("great", "excellent", "love", "amazing")
#   - negative_words: count of negative words ("bad", "terrible", "hate", "awful")
#   - sentiment_score: positive_words - negative_words
# Hint: Use str_count() to count pattern occurrences

feedback_clean <- feedback_clean %>%
  mutate(
    # Your code here:
    positive_words = str_count(feedback_clean, "great|excellent|love|amazing"),
    negative_words = str_count(feedback_clean, "bad|terrible|Hate|awful"),
    sentiment_score = positive_words - negative_words
  )

# Display sentiment analysis results
cat("Sentiment Analysis Results:\n")
feedback_clean %>%
  select(feedback_clean, positive_words, negative_words, sentiment_score) %>%
  head(10) %>%
  print()

# Summary
cat("\nOverall Sentiment Summary:\n")
cat("Average sentiment score:", mean(feedback_clean$sentiment_score), "\n")
cat("Positive reviews:", sum(feedback_clean$sentiment_score > 0), "\n")
cat("Negative reviews:", sum(feedback_clean$sentiment_score < 0), "\n")

Sentiment Analysis Results:
                    feedback_clean positive_words negative_words
1       highly recommend this item              0              0
2                excellent service              1              0
3             poor quality control              0              0
4  average product nothing special              0              0
5         amazing customer support              1              0
6         amazing customer support              1              0
7  average product nothing special              0              0
8             good value for money              0              0
9       highly recommend this item              0              0
10      highly recommend this item              0              0
   sentiment_score
1                0
2                1
3                0
4                0
5                1
6                1
7                0
8                0
9                0
10               0

Overall Sentiment Summary:
Average sentiment sc

## Part 4: Date Parsing and Component Extraction

**Business Context:** Transaction dates need to be parsed and analyzed to understand customer behavior patterns.

**Your Tasks:**
1. Parse transaction dates from text to Date objects
2. Extract date components (year, month, day, weekday)
3. Identify weekend vs weekday transactions
4. Extract quarter and month names

**Key Functions:** `ymd()`, `mdy()`, `dmy()`, `year()`, `month()`, `day()`, `wday()`, `quarter()`

In [12]:
# Task 4.1: Parse Transaction Dates
# TODO: Create a new column 'date_parsed' that parses the transaction_date column
# Hint: Check the format of transaction_date first, then use ymd(), mdy(), or dmy()

transactions_clean <- transactions %>%
  mutate(
    # Your code here:
    date_parsed = case_when(
      !is.na(mdy_hm(Transaction_DateTime)) ~ mdy_hm(Transaction_DateTime),
      !is.na(ymd_hm(Transaction_DateTime)) ~ ymd_hm(Transaction_DateTime),
      !is.na(dmy_hm(Transaction_DateTime)) ~ dmy_hm(Transaction_DateTime),
    )
  )

# Verify parsing worked
cat("Date Parsing Results:\n")
transactions_clean %>%
  select(Transaction_DateTime, date_parsed) %>%
  head(10) %>%
  print()
  unparsed_count <- sum(is.na(transactions_clean$date_parsed))
  cat("\nNumber of unparsed dates:", unparsed_count, "\n")

Warning message:
“There were 6 warnings in `mutate()`.
The first warning was:
ℹ In argument: `date_parsed = case_when(...)`.
Caused by warning:
!  62 failed to parse.
ℹ Run `dplyr::last_dplyr_warnings()` to see the 5 remaining warnings.”


Date Parsing Results:
   Transaction_DateTime         date_parsed
1          4/5/24 14:30 2024-04-05 14:30:00
2        3/15/241 14:30                <NA>
3         3/15/24 14:30 2024-03-15 14:30:00
4          3/20/24 9:15 2024-03-20 09:15:00
5          3/20/24 9:15 2024-03-20 09:15:00
6          3/20/24 9:15 2024-03-20 09:15:00
7          3/20/24 9:15 2024-03-20 09:15:00
8         3/15/24 14:30 2024-03-15 14:30:00
9   25-03-2024 16:45:30                <NA>
10         4/5/24 14:30 2024-04-05 14:30:00

Number of unparsed dates: 62 


In [13]:
# Task 4.2: Extract Date Components
# TODO: Create the following new columns:
#   - trans_year: Extract year from date_parsed
#   - trans_month: Extract month number from date_parsed
#   - trans_month_name: Extract month name (use label=TRUE, abbr=FALSE)
#   - trans_day: Extract day of month from date_parsed
#   - trans_weekday: Extract weekday name (use label=TRUE, abbr=FALSE)
#   - trans_quarter: Extract quarter from date_parsed

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    trans_year = year(date_parsed),
    trans_month = month(date_parsed),
    trans_month_name = month(date_parsed, label=TRUE, abbr=FALSE),
    trans_day = day(date_parsed),
    trans_weekday = wday(date_parsed, label=TRUE, abbr=FALSE),
    trans_quarter = quarter(date_parsed)
  )

# Display results
cat("Date Component Extraction:\n")
transactions_clean %>%
  select(date_parsed, trans_month_name, trans_weekday, trans_quarter) %>%
  head(10) %>%
  print()

Date Component Extraction:
           date_parsed trans_month_name trans_weekday trans_quarter
1  2024-04-05 14:30:00            April        Friday             2
2                 <NA>             <NA>          <NA>            NA
3  2024-03-15 14:30:00            March        Friday             1
4  2024-03-20 09:15:00            March     Wednesday             1
5  2024-03-20 09:15:00            March     Wednesday             1
6  2024-03-20 09:15:00            March     Wednesday             1
7  2024-03-20 09:15:00            March     Wednesday             1
8  2024-03-15 14:30:00            March        Friday             1
9                 <NA>             <NA>          <NA>            NA
10 2024-04-05 14:30:00            April        Friday             2


In [14]:
# Task 4.3: Identify Weekend Transactions
# TODO: Create a new column 'is_weekend' that is TRUE if the transaction was on Saturday or Sunday
# Hint: Use wday() which returns 1 for Sunday and 7 for Saturday
# Hint: Use %in% c(1, 7) to check if day is weekend

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    is_weekend = wday(date_parsed) %in% c(1, 7)
  )

# Summary
cat("Weekend vs Weekday Transactions:\n")
table(transactions_clean$is_weekend) %>% print()

cat("\nPercentage of weekend transactions:",
    round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1), "%\n")

Weekend vs Weekday Transactions:

FALSE 
  150 

Percentage of weekend transactions: 0 %


## Part 5: Date Calculations and Customer Recency Analysis

**Business Context:** Understanding how recently customers transacted helps identify at-risk customers for re-engagement campaigns.

**Your Tasks:**
1. Calculate days since each transaction
2. Categorize customers by recency (Recent, Moderate, Old)
3. Identify customers who haven't transacted in 90+ days
4. Calculate average days between transactions per customer

**Key Functions:** `today()`, date arithmetic, `case_when()`

In [15]:
# Task 5.1: Calculate Days Since Transaction
# TODO: Create a new column 'days_since' that calculates days from date_parsed to today()
# Hint: Use as.numeric(today() - date_parsed)

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    days_since = as.numeric(today() - as.Date(date_parsed))
  )

# Display results
cat("Days Since Transaction:\n")
transactions_clean %>%
  select(CustomerID, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  head(10) %>%
  print()

Days Since Transaction:
   CustomerID         date_parsed days_since
1          12 2024-03-15 14:30:00        577
2          30 2024-03-15 14:30:00        577
3          45 2024-03-15 14:30:00        577
4           2 2024-03-15 14:30:00        577
5          18 2024-03-15 14:30:00        577
6          34 2024-03-15 14:30:00        577
7          48 2024-03-15 14:30:00        577
8          28 2024-03-15 14:30:00        577
9          30 2024-03-15 14:30:00        577
10         33 2024-03-15 14:30:00        577


In [30]:
# Task 5.2: Categorize by Recency
# TODO: Create a new column 'recency_category' using case_when():
#   - "Recent" if days_since <= 30
#   - "Moderate" if days_since <= 90
#   - "At Risk" if days_since > 90

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    recency_category = case_when(
      days_since <= 30  ~ "Recent"),
     recency_category = case_when(days_since <= 90 ~ "Moderate"),
     recency_category = case_when(days_since > 90 ~ "At Risk")
    
  )

# Display distribution
cat("Recency Category Distribution:\n")
table(transactions_clean$recency_category) %>% print()

# Show at-risk customers
cat("\nAt-Risk Customers (>90 days):\n")
transactions_clean %>%
  filter(recency_category == "At Risk") %>%
  select(CustomerID, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  print()

Recency Category Distribution:

At Risk 
     88 

At-Risk Customers (>90 days):
   CustomerID         date_parsed days_since
1          12 2024-03-15 14:30:00        577
2          30 2024-03-15 14:30:00        577
3          45 2024-03-15 14:30:00        577
4           2 2024-03-15 14:30:00        577
5          18 2024-03-15 14:30:00        577
6          34 2024-03-15 14:30:00        577
7          48 2024-03-15 14:30:00        577
8          28 2024-03-15 14:30:00        577
9          30 2024-03-15 14:30:00        577
10         33 2024-03-15 14:30:00        577
11         11 2024-03-15 14:30:00        577
12         36 2024-03-15 14:30:00        577
13          6 2024-03-15 14:30:00        577
14          8 2024-03-15 14:30:00        577
15         38 2024-03-15 14:30:00        577
16         49 2024-03-15 14:30:00        577
17         28 2024-03-15 14:30:00        577
18         44 2024-03-15 14:30:00        577
19         33 2024-03-15 14:30:00        577
20         29 2024-

## Part 6: Combined String and Date Operations

**Business Context:** Create personalized customer outreach messages based on purchase recency.

**Your Tasks:**
1. Extract first names from customer names
2. Create personalized messages based on recency
3. Analyze transaction patterns by weekday
4. Identify best customers (recent + high value)

**Key Functions:** Combine `str_extract()`, date calculations, `case_when()`, `group_by()`, `summarize()`

In [17]:
# Task 6.1: Extract First Names and Create Personalized Messages
# TODO: Create two new columns:
#   - first_name: Extract first name from customer_name (everything before first space)
#   - personalized_message: Create message based on recency_category
#     * Recent: "Hi [name]! Thanks for your recent purchase!"
#     * Moderate: "Hi [name], we miss you! Check out our new products."
#     * At Risk: "Hi [name], it's been a while! Here's a special offer for you."
# Hint: Use str_extract() with pattern "^\\\\w+" for first name
# Hint: Use paste() to combine strings in case_when()

customer_outreach <- transactions_clean %>%
  mutate(
    # Your code here:
    first_name = str_extract(CustomerID, "^\\w+"),
    personalized_message = case_when(
      recency_category == "Recent" ~ paste("Hi", first_name, "! Thanks for your recent purchase!"),
      recency_category == "Moderate" ~ paste("Hi", first_name, ", we miss you! Check out our new products."),
      recency_category == "At Risk" ~ paste("Hi", first_name, ", it's been a while! Here's a special offer for you.")
    )
  )

# Display personalized messages
cat("Personalized Customer Messages:\n")
customer_outreach %>%
  select(CustomerID, first_name, days_since, personalized_message) %>%
  head(10) %>%
  print()

Personalized Customer Messages:
   CustomerID first_name days_since
1          26         26        556
2          21         21         NA
3          12         12        577
4           6          6        572
5          32         32        572
6          27         27        572
7          31         31        572
8          30         30        577
9          31         31         NA
10         13         13        556
                                         personalized_message
1  Hi 26 , it's been a while! Here's a special offer for you.
2                                                        <NA>
3  Hi 12 , it's been a while! Here's a special offer for you.
4   Hi 6 , it's been a while! Here's a special offer for you.
5  Hi 32 , it's been a while! Here's a special offer for you.
6  Hi 27 , it's been a while! Here's a special offer for you.
7  Hi 31 , it's been a while! Here's a special offer for you.
8  Hi 30 , it's been a while! Here's a special offer for you.
9             

In [25]:
# Task 6.2: Analyze Transaction Patterns by Weekday
# TODO: Group by trans_weekday and calculate:
#   - transaction_count: number of transactions
#   - total_amount: sum of amount (if available)
#   - avg_amount: average amount per transaction
# TODO: Arrange by transaction_count descending
names(transactions_clean)
weekday_patterns <- transactions_clean %>%
  # Your code here:
  group_by(trans_weekday) %>%
  transaction_count = n() %>%
  total_amount = sum(Amount, na.rm=TRUE) %>%
  avg_amount = mean(Amount, na.rm=TRUE) %>%
  arrange(desc(transaction_count))

# Display results
cat("Transaction Patterns by Weekday:\n")
print(weekday_patterns)

# Identify busiest day
busiest_day <- weekday_patterns$trans_weekday[1]
cat("\n🔥 Busiest day:", as.character(busiest_day), "\n")

[1] "LogID"                "CustomerID"           "Transaction_DateTime"
 [4] "Amount"               "Status"               "date_parsed"         
 [7] "trans_year"           "trans_month"          "trans_month_name"    
[10] "trans_day"            "trans_weekday"        "trans_quarter"       
[13] "is_weekend"           "days_since"           "recency_category"

ERROR: Error: object 'Amount' not found


In [24]:
# Task 6.3: Monthly Transaction Analysis
# TODO: Group by trans_month_name and calculate:
#   - transaction_count
#   - unique_customers: use n_distinct(customer_name)
# TODO: Arrange by trans_month (to show chronological order)

monthly_patterns <- transactions_clean %>%
  # Your code here:
  group_by(trans_month_name) %>%
  transaction_count = n() %>%
  unique_customers = n_distinct(CustomerID) %>%
  arrange(desc(trans_month))

# Display results
cat("Monthly Transaction Patterns:\n")
print(monthly_patterns)

ERROR: Error: object 'CustomerID' not found


## Part 7: Business Intelligence Summary

**Business Context:** Create an executive summary that combines all your analyses into actionable insights.

**Your Tasks:**
1. Calculate key metrics across all datasets
2. Identify top products and categories
3. Summarize customer sentiment
4. Provide data-driven recommendations

In [ ]:
# Task 7.1: Create Business Intelligence Dashboard

cat("\n", rep("=", 60), "\n")
cat("         BUSINESS INTELLIGENCE SUMMARY\n")
cat(rep("=", 60), "\n\n")

# Product Analysis
cat("📦 PRODUCT ANALYSIS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total number of products
#   - Number of wireless products
#   - Number of premium products
#   - Most common category


# Customer Sentiment
cat("\n💬 CUSTOMER SENTIMENT\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total feedback entries
#   - Average sentiment score
#   - Percentage of positive reviews
#   - Percentage of negative reviews


# Transaction Patterns
cat("\n📊 TRANSACTION PATTERNS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total transactions
#   - Date range (earliest to latest)
#   - Busiest weekday
#   - Weekend transaction percentage


# Customer Recency
cat("\n👥 CUSTOMER RECENCY\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Number of recent customers (< 30 days)
#   - Number of at-risk customers (> 90 days)
#   - Percentage needing re-engagement



In [ ]:
# Task 7.2: Identify Top Products by Category
# TODO: Group products by category_clean and count products in each
# TODO: Arrange by count descending
# TODO: Display top 5 categories

top_categories <- products_clean %>%
  # Your code here:
  

cat("Top Product Categories:\n")
print(top_categories)

## Part 8: Reflection Questions

Answer the following questions based on your analysis. Write your answers in the markdown cells below.

### Question 8.1: Data Quality Impact

**How did cleaning the text data (removing spaces, standardizing case) improve your ability to analyze the data? Provide specific examples from your homework.**

Your answer here: It becomes a lot easier to analyze when the data it standarized. If I had to deal with customers typos and inconsistent way of writing messages it would be more time consuming and difficult to find the data I need.



### Question 8.2: Pattern Detection Value

**What business insights did you gain from detecting patterns in product names (wireless, premium, gaming)? How could a business use this information?**

Your answer here: Businesses can use that information to see what their lowest performing products are and try to market that product better. They cna also see which products people are the most pleased with and figure out which ones they have to improve development on.



### Question 8.3: Date Analysis Importance

**Why is analyzing transaction dates by weekday and month important for business operations? Provide at least three specific business applications.**

Your answer here:

1. 
2. 
3. 


### Question 8.4: Customer Recency Strategy

**Based on your recency analysis, what specific actions would you recommend for customers in each category (Recent, Moderate, At Risk)? How would you prioritize these actions?**

Your answer here: For the at risk customers I would reach out and inquire why they hadn't purchased anything recently. I would also offer them a slight discount for their next order. The moderate customers I would inquire just like with the at risk customers but I would offer a a discount on the last item they ordered and try to entice them to order more frequently. The recent customers I would show my appreciation too by sending a thank you menu and offering something such as free shipping on their next order or sending some sort of small free merchandise with their next order.



### Question 8.5: Sentiment Analysis Application

**How could the sentiment analysis you performed be used to improve products or customer service? What are the limitations of this simple sentiment analysis approach?**

Your answer here:



### Question 8.6: Real-World Application

**Describe a real business scenario where you would need to combine string manipulation and date analysis (like you did in this homework). What insights would you be trying to discover?**

Your answer here:



## Summary and Submission

### What You've Accomplished

In this homework, you've successfully:
- ✅ Cleaned and standardized messy text data using `stringr` functions
- ✅ Detected patterns and extracted information from text
- ✅ Parsed dates and extracted temporal components using `lubridate`
- ✅ Calculated customer recency for segmentation
- ✅ Analyzed transaction patterns by time periods
- ✅ Combined string and date operations for business insights
- ✅ Created personalized customer communications
- ✅ Generated executive-ready business intelligence summaries

### Key Skills Mastered

**String Manipulation:**
- `str_trim()`, `str_squish()` - Whitespace handling
- `str_to_lower()`, `str_to_upper()`, `str_to_title()` - Case conversion
- `str_detect()` - Pattern detection
- `str_extract()` - Information extraction
- `str_count()` - Pattern counting

**Date/Time Operations:**
- `ymd()`, `mdy()`, `dmy()` - Date parsing
- `year()`, `month()`, `day()`, `wday()` - Component extraction
- `quarter()` - Period extraction
- `today()` - Current date
- Date arithmetic - Calculating differences

**Business Applications:**
- Data cleaning and standardization
- Customer segmentation by recency
- Sentiment analysis
- Pattern identification
- Temporal trend analysis
- Personalized communication

### Submission Checklist

Before submitting, ensure you have:
- [ ] Entered your name, student ID, and date at the top
- [ ] Completed all code tasks (Parts 1-7)
- [ ] Run all cells successfully without errors
- [ ] Answered all reflection questions (Part 8)
- [ ] Used proper commenting in your code
- [ ] Used the pipe operator (`%>%`) where appropriate
- [ ] Verified your results make business sense
- [ ] Checked for any remaining TODO comments

### Grading Criteria

Your homework will be evaluated on:
- **Code Correctness (40%)**: All tasks completed correctly
- **Code Quality (20%)**: Clean, well-commented, efficient code
- **Business Understanding (20%)**: Demonstrates understanding of business applications
- **Reflection Questions (15%)**: Thoughtful, complete answers
- **Presentation (5%)**: Professional formatting and organization

### Next Steps

In Lesson 8, you'll learn:
- Advanced data wrangling with complex pipelines
- Sophisticated conditional logic with `case_when()`
- Data validation and quality checks
- Creating reproducible analysis workflows
- Professional best practices for business analytics

**Great work on completing this assignment! 🎉**